enterprise/server/remote_execution/copy_on_write/copy_on_write.go

```diff
@@ -473,6 +516,73 @@ func (s *COWStore) initDirtyChunk(offset int64, size int64) (ogChunk *Mmap, newC
 	return ogChunk, newChunk, nil
 }
 
+type eagerFetchData struct {
+	// Chunk offset to eagerly fetch
+	offset int64
+}
+
+func (s *COWStore) eagerFetchNextChunks(offset int64) {
+	currentOffset := offset
+	for i := 0; i < numChunksToEagerFetch; i++ {
+		s.sendNonBlockingEagerFetch(currentOffset)
+		currentOffset += s.chunkSizeBytes
+	}
+}
+
+func (s *COWStore) sendNonBlockingEagerFetch(offset int64) {
+	select {
+	case s.eagerFetchChan <- &eagerFetchData{offset: offset}:
+	default:
+		log.Debug("COWStore eager fetch chan full")
+	}
+}
+
+func (s *COWStore) eagerFetchChunksInBackground() {
+	rateLimiter := rate.NewLimiter(rate.Limit(maxEagerFetchesPerSec), 1)
+	eg := &errgroup.Group{}
+	eg.SetLimit(eagerFetchConcurrency)
+
+	for {
+		select {
+		case <-s.quitChan:
+			_ = eg.Wait()
+			return
+		case d := <-s.eagerFetchChan:
+			if err := rateLimiter.Wait(s.ctx); err != nil {
+				if err != nil {
+					log.Errorf("COWStore eager fetch rate limiter failed, stopping eager fetches: %s", err)
+				}
+				_ = eg.Wait()
+				return
+			}
+			eg.Go(func() error {
+				err := s.fetchChunk(d.offset)
+				if err != nil {
+					log.Warningf("COWStore eager fetch chunk failed with: %s", err)
+				}
+				return nil
+			})
+		}
+	}
+}
+
+func (s *COWStore) fetchChunk(offset int64) error {
+	s.mu.RLock()
+	c := s.chunks[offset]
+	s.mu.RUnlock()
+
+	// Skip holes
+	if c == nil {
+		return nil
+	}
+
```

Comment:
nit: maybe move the rateLimiter.Wait() after this line?

Reply:
I'm wondering if the rate limiter could also help protect over-locking on `chunks`. I'm just going to leave it as is for now - probably wouldn't be a big difference either way